File: approach-2-reactive/src/main/java/com/example/approach_2_reactive/controller/HelloWorldController.java
package com.example.approach_2_reactive.controller;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.time.Instant;
import java.util.concurrent.atomic.AtomicInteger;

@RestController
public class HelloWorldController {

    private final AtomicInteger counter = new AtomicInteger(0);

    private final WebClient webClient;

    public HelloWorldController(WebClient webClient) {
        this.webClient = webClient;
    }

    @GetMapping("/hello-world")
    public Mono<String> helloWorld() {
        System.out.printf("%s: %s: REQUEST#: %d\n", Thread.currentThread(), Instant.now(), counter.addAndGet(1));

        return webClient.get()
                .uri("http://localhost:9091/hello-slow")
                .retrieve()
                .bodyToMono(String.class);
    }

}

File: approach-3-kotlin-coroutine/src/main/kotlin/com/example/approach_3_kotlin_coroutine/controller/HelloWorldController.kt
package com.example.approach_3_kotlin_coroutine.controller

import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBodyOrNull
import java.time.Instant
import java.util.concurrent.atomic.AtomicInteger

@RestController
class HelloWorldController(private val webClient: WebClient) {
    private val counter = AtomicInteger(0)

    @GetMapping("/hello-world")
    suspend fun helloWorld(): String? {
        println("${Thread.currentThread()}: ${Instant.now()}: REQUEST#: ${counter.addAndGet(1)}")
        return webClient.get()
            .uri("http://localhost:9091/hello-slow")
            .retrieve()
            .awaitBodyOrNull()
    }
}
